draft-ietf-quic-load-balancers.md

```diff
@@ -450,45 +450,48 @@ length, splitting an odd octet in half if necessary. For example,
 0x7040b81b55ccf3 would split into a left_0 of 0x7040b81 and right_0 of
 0xb55ccf3.
 
-3. Encrypt the result of expand_left(left_0) to obtain a ciphertext.
+3. Encrypt the result of expand_left(left_0, index)) to obtain a ciphertext,
+where 'index' is one octet: the two most significant bits of which are 0b00,
+and the six least significant bits are the length of the resulting connection
+ID in bytes, cid_len.
 
 4. XOR the least significant bits of the ciphertext with right_0 to form
 right_1.
 
     Thus steps 3 and 4 can be expressed as
     ```
     right_1 = right_0 ^ truncate_right(
-                            AES_ECB(key, expand_left(left_0, 0x01)),
+                            AES_ECB(key, expand_left(left_0, cid_len)),
                             len(right_0))
     ```
 
 5. Repeat steps 3 and 4, but use them to compute left_1 by expanding and
-encrypting right_1 with the most significant octet as 0x02 and XOR the
-results with left_0.
+encrypting right_1 with the most significant octet as the concatenation of
+0b01 and cid_len, and XOR the results with left_0.
 
     ```
     left_1 = left_0 ^ truncate_left(
-                          AES_ECB(key, expand_right(right_1, 0x02)),
+                          AES_ECB(key, expand_right(right_1, 0x40 | cid_len)),
                           len(left_0))
     ```
 
 6. Repeat steps 3 and 4, but use them to compute right_2 by expanding and
-encrypting left_1 with the least significant octet as 0x03 and XOR the
-results with right_1.
+encrypting left_1 with the least significant octet as the concatenation of
+0b10 and cid_len, and XOR the results with right_1.
 
     ```
     right_2 = right_1 ^ truncate_right(
-                            AES_ECB(key, expand_left(left_1, 0x03)),
+                            AES_ECB(key, expand_left(left_1, 0x80 | cid_len)),
                             len(right_1))
     ```
 
 7. Repeat steps 3 and 4, but use them to compute left_2 by expanding and
-encrypting right_2 with the most significant octet as 0x04 and XOR the
-results with left_1.
+encrypting right_2 with the most significant octet as the concatenation of
+0b11 ands cid_len, and XOR the results with left_1.
 
     ```
     left_2 = left_1 ^ truncate_left(
-                          AES_ECB(key, expand_right(right_2, 0x04)),
+                          AES_ECB(key, expand_right(right_2, 0xc0 | cid_len)),
                           len(left_1))
     ```
```

Comment:
So the index is effectively:

* 3 bits for the identifier of the step,
* 5 bits for the length of the CID.

If we stick with QUIC v1, the max CID length is 20 bytes. 5 bits definitely suffice. But if we ever increase the number of steps to 12, 3 bits will not suffice.

Do we really need to pack bits? The "right" and "left" fields are at most 10 bytes, so we have 6 bytes available for the "tweak". Can't we have 1 byte for the step number and one byte for the CID length? 
 The "right" and "left" can be at most 10 bytes.

Reply:
This comment is addressed in the new text.